src/internal/components/pluggableVisualizations/xirr/tests/xirrBucketHelper.spec.ts

```diff
@@ -0,0 +1,573 @@
+// (C) 2019 GoodData Corporation
+import { getXirrBuckets } from "../xirrBucketHelper";
+import { IReferencePoint, IBucket } from "../../../../interfaces/Visualization";
+import * as BucketNames from "../../../../../constants/bucketNames";
+import * as referencePointMocks from "../../../../mocks/referencePointMocks";
+
+describe("xirrBucketHelper", () => {
+    const validMeasureBucket: IBucket = {
+        localIdentifier: BucketNames.MEASURES,
+        items: [referencePointMocks.masterMeasureItems[0]],
+    };
+
+    const emptyMeasureBucket: IBucket = {
+        localIdentifier: BucketNames.MEASURES,
+        items: [],
+    };
+
+    const validAttributeBucket: IBucket = {
+        localIdentifier: BucketNames.ATTRIBUTE,
+        items: [referencePointMocks.dateItem],
+    };
+
+    const emptyAttributeBucket: IBucket = {
+        localIdentifier: BucketNames.ATTRIBUTE,
+        items: [],
+    };
+
+    const getReferencePoint = (buckets: IBucket[]): IReferencePoint => ({
+        buckets,
+        filters: {
+            items: [],
+            localIdentifier: "filters",
+        },
+    });
+
+    describe("getXirrBuckets", () => {
+        it("should maintain existing buckets", () => {
+            const buckets = [validMeasureBucket, validAttributeBucket];
+            const input = getReferencePoint(buckets);
+            const actual = getXirrBuckets(input);
+
+            expect(actual).toEqual(buckets);
+        });
+
+        it("should add default buckets if some are missing", () => {
+            const buckets = [validAttributeBucket];
+            const input = getReferencePoint(buckets);
+            const actual = getXirrBuckets(input);
+
+            expect(actual).toEqual([emptyMeasureBucket, validAttributeBucket]);
+        });
+
+        it("should remove extraneous buckets", () => {
+            const buckets = [
+                {
+                    ...validMeasureBucket,
+                    items: [...validMeasureBucket.items, referencePointMocks.masterMeasureItems[1]],
+                },
+                validAttributeBucket,
+            ];
+            const input = getReferencePoint(buckets);
+            const actual = getXirrBuckets(input);
+
+            expect(actual).toEqual([validMeasureBucket, validAttributeBucket]);
+        });
+
+        type MigrationTestCase = [string, IBucket[], IBucket[]];
+        const migrationTestCases: MigrationTestCase[] = [
+            [
+                "table without date attribute",
+                [
+                    {
+                        localIdentifier: BucketNames.MEASURES,
+                        items: [referencePointMocks.masterMeasureItems[0]],
+                    },
```

Comment:
These buckets are repeated many times. Might shorten the code if extracted.

Reply:
Good point, changed, thanks!